### PR TITLE
Extend ambiguous definitions check to include casgns

### PR DESCRIPTION
**Pull request by aadi-stripe**
Extends the check for "ambiguous definitions" in the resolver to also include casgns.

### Motivation

The ambiguous definitions [check](https://sorbet.org/docs/error-reference#5068) was added to prevent namespace conflict/ambiguity in the Stripe codebase at runtime. There was a gap in the enforcement -- in that it did not prevent casgns from being ambiguously namespaced. This change closes that gap.

### Test plan

See included automated tests. Tested on Stripe codebase.

### CODE REVIEW

**Comment by aadi-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_M7VtvouJGrGkj7
→ https://go/builds/bui_M7VtURuiB1FoDo

File: resolver/resolver.cc [hunk 1]
```diff
@@ -1298,7 +1298,7 @@ class ResolveConstantsWalk {
         nesting_ = nesting_->parent;
     }
 
-    const bool checkAmbiguousDefinition(core::Context ctx) {
+    const bool checkAmbiguousDefinition(core::Context ctx, core::SymbolRef curSym, shared_ptr<Nesting> curNesting) {
```
Comment (froydnj) [inline resolver/resolver.cc:1301]: Can we take the `shared_ptr` by reference here so we're not constantly doing atomic reference counts unnecessarily?

```suggestion
    const bool checkAmbiguousDefinition(core::Context ctx, core::SymbolRef curSym, const shared_ptr<Nesting> &curNesting) {
```

File: resolver/resolver.cc [hunk 2]
```diff
@@ -1309,24 +1309,23 @@ class ResolveConstantsWalk {
...
+        if (curOwner == curNesting->scope) {
             return false;
         }
 
         return true;
     }
 
-    const core::SymbolRef findAnyDefinitionAmbiguousWithCurrent(core::Context ctx) {
+    const core::SymbolRef findAnyDefinitionAmbiguousWithCurrent(core::Context ctx, core::SymbolRef curSym,
+                                                                shared_ptr<Nesting> curNesting) {
```
Comment (froydnj) [inline resolver/resolver.cc:1327]: Same comment here, though I think taking by `const &` means that we need to rewrite the loop in this function a little further down...which is probably not a bad idea anyway.

```suggestion
    const core::SymbolRef findAnyDefinitionAmbiguousWithCurrent(core::Context ctx, core::SymbolRef curSym,
                                                                const shared_ptr<Nesting> &curNesting) {
```

**Review by froydnj — APPROVED** [review]
Up to you whether you want to fix these here or in a followup; since this code happens during a parallel phase, the efficiency matters slightly less, but even in parallel code serial efficiency still matters.

**Comment by aadi-stripe** [comment]
@froydnj fixed

**Review by froydnj — APPROVED** [review]

**Pull request merged**